### PR TITLE
Add error boundary components and exception logging

### LINKED ISSUE #1187: Improve integration with sentry
We use [Sentry](https://sentry.io) for crash reporting in the apps, and they have a [react-native client](https://github.com/getsentry/sentry-react-native) that allows for better stack traces. They even support mixed stack traces on iOS.

image

I did some initial experiments a while back but I recall the initialization of the JS client was running into some issues because we already had a client initialized on the iOS side.

We need more research to see how we can make this work.

### PR DESCRIPTION
**Pull request by fluiddot**
**Related PRs:**
* https://github.com/WordPress/gutenberg/pull/59221
* https://github.com/wordpress-mobile/WordPress-iOS/pull/22655
* https://github.com/Automattic/Automattic-Tracks-iOS/pull/278
* https://github.com/wordpress-mobile/WordPress-Android/pull/20359
* https://github.com/Automattic/Automattic-Tracks-Android/pull/202

Closes https://github.com/wordpress-mobile/gutenberg-mobile/issues/1187#issuecomment-508570299.

The changes coming from Gutenberg enable the error boundary at the editor and block level. As well as provide a bridge function to log JavaScript exceptions to the Crash Logging service used in the host apps. 

Additionally, this PR includes the following changes:
* Set a global error handler to log exceptions caused by unhandled errors.
* Include the source map files as Buildkite artifacts. Previously, it was limited to jobs triggered by Git tags.
* Include the source map files in iOS XCFramework so they can be uploaded to the Crash logging service to symbolicate the stack trace.

## To test
Since it involves testing exceptions, we need to modify the code to force errors and generate an installable build. A test PR has been created for this purpose, follow the testing instructions from https://github.com/wordpress-mobile/gutenberg-mobile/pull/6654.

PR submission checklist:

- [ ] I have considered adding unit tests where possible.
- [x] I have considered if this change warrants user-facing release notes [more info](https://github.com/wordpress-mobile/gutenberg-mobile/blob/trunk/Release-notes.md) and have added them to [RELEASE-NOTES.txt](https://github.com/wordpress-mobile/gutenberg-mobile/blob/trunk/RELEASE-NOTES.txt) if necessary.

### CODE REVIEW

File: .buildkite/pipeline.yml [hunk 1]
```diff
@@ -143,29 +143,19 @@ steps:
         echo "--- :android: Build Android bundle"
         npm run bundle:android
 
-        echo "--- :arrow_up: Upload Android bundle artifact"
+        echo "--- :arrow_up: Upload Android bundle and source map artifacts"
         buildkite-agent artifact upload bundle/android/App.js
-
-        if [[ -n "$BUILDKITE_TAG" ]]; then
-          echo "--- :arrow_up: Upload Android source map"
-          node gutenberg/node_modules/react-native/scripts/compose-source-maps.js bundle/android/App.text.js.map bundle/android/App.js.map -o bundle/android/App.composed.js.map
```
Comment (fluiddot) [inline .buildkite/pipeline.yml:151]: The composed source map is now generated as part of the `bundle` npm command.

File: src/errorLogging/index.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+					'Encountered multiple fatals in a row. The latest:',
+					error
+				);
+				return;
+			}
+			handlingFatal = true;
+		}
+
+		// TODO: Set severity and exception mechanism.
+		// https://github.com/getsentry/sentry-react-native/blob/adfb66f16438dfd98f280307844778c7291b584b/src/js/integrations/reactnativeerrorhandlers.ts#L235-L239
```
Comment (fluiddot) [inline src/errorLogging/index.js]: The severity and exception mechanism should be added in the [`logException` function](https://github.com/WordPress/gutenberg/blob/12aff798a3c3f0fb598f520a7901b86ed337cb9c/packages/react-native-bridge/index.js#L537-L562) when passing a parameter (e.g. `unhandled`).

Comment (fluiddot) [reply src/errorLogging/index.js]: The mechanism is inserted in the Crash logging service ([reference](https://github.com/Automattic/Automattic-Tracks-iOS/blob/decdafc1c8b08ff83935cccc8880b0809ca9e5c6/Sources/Remote%20Logging/Crash%20Logging/SentryEventJSException.swift#L65-L68)).

**Review by jhnstn — APPROVED** [review]
The changes look good and the iOS test build works as expected.

I know we still need to test with an Android build but unless anything changes, this looks ready to go.

**Pull request merged**